File: daily/2024/2807_Insert_Greatest_Common_Divisors_in_Linked_List.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* create_list(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    ListNode* head = new ListNode(nodes[0]);
    ListNode* curr = head;
    for (int i = 1; i < nodes.size(); i++){
        curr->next = new ListNode(nodes[i]);
        curr = curr->next;
    }
    return head;
}

void print_list(ListNode* head){
    while(head){
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;
    return;
}

class Solution {
public:
    int GCD(int a, int b){
        if (a % b == 0) return b;
        return GCD(b, a % b);
    }
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* prev = head;
        ListNode* curr = head->next;
        while(curr){
            prev->next = new ListNode(GCD(prev->val, curr->val));
            prev->next->next = curr;
            prev = curr;
            curr = curr->next;
        }
        return head;
    }
};

int main(){
    vector<int> nodes = {18,6,10,3};
    // vector<int> nodes = {7};
    ListNode* head = create_list(nodes);
    Solution S;
    S.insertGreatestCommonDivisors(head);
    print_list(head);
    return 0;
}

File: daily_python/2807_Insert_Greatest_Common_Divisors_in_Linked_List.py
import os
import sys
from typing import List, Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
def create_list(nodes: List[int]) -> ListNode :
    if not nodes:
        return None
    head = ListNode(nodes[0])
    curr = head
    for i in range(1, len(nodes)):
        curr.next = ListNode(nodes[i])
        curr = curr.next
    return head

def print_list(head: ListNode) -> None:
    while(head):
        print(head.val, end = ' ')
        head = head.next
    print()

class Solution:
    def GCD(self, a: int, b: int) -> int:
        if a % b == 0:
            return b
        return self.GCD(b, a % b)
    
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = head
        curr = head.next
        while curr:
            prev.next = ListNode(self.GCD(prev.val, curr.val))
            prev.next.next = curr
            prev = curr
            curr = prev.next
        return head
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            head = create_list(nodes)
            dataset.append(head)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for head in dataset:
        results.append(solution.insertGreatestCommonDivisors(head))
    for index, result in enumerate(results):
        print(f'Example : ', end = '')
        print_list(result)